### PR TITLE
[None][fix] Re-plan FlashInfer decode schedule on page-table changes under CUDA graphs

### PR DESCRIPTION
**Pull request by brnguyen2**
## Description

With CUDA graphs enabled, the FlashInfer decode kernel runs the split-KV tile schedule that `plan()` computed: the request/kv-tile index lists and `kv_chunk_size` live in the wrapper's persistent plan workspace, and graph replay never re-enters `forward_impl`. Whatever plan was captured is what every replay executes, so the plan has to be kept current from `prepare()`.

`_refresh_fa2_cuda_graph_plans()` already does that, keyed on a page tuple (`fa2_plan_num_blocks`) recorded when the decode plan ran. The recording was gated on the wrapper being both a graph tensor-core wrapper (`head_dim > 128`) *and* labelled `'fa2'`. Any decode wrapper outside that case never got a tuple, so the refresh skipped it.

That is invisible for a model with a single decode wrapper, because `prepare()` plans it eagerly. It is not invisible for a model with more than one decode wrapper — a mixed sliding-window/full-attention checkpoint has one wrapper per distinct window — because `prepare()` defers multi-wrapper plans to `forward_impl`, which graph replay never reaches. Every replay then reused the plan captured for the warmup batch: one `max_seq_len` dummy in generation slot 0 and one-token dummies elsewhere. Slot 0 kept full KV coverage, so batch size 1 looked fine, but every other slot attended only its first `kv_chunk_size` tokens and never its own output, and batched decode degenerated into repetition. The TRTLLM attention backend was unaffected.

The fix records the page tuple for every non-`trtllm-gen` decode wrapper whenever the metadata is graph-captured, so the refresh re-plans all of them when the generation page counts change. `trtllm-gen` stays excluded because it already refreshes its own block tables and KV lengths per step in `prepare()`; the eager path is unchanged, since `is_cuda_graph` is false there and `forward_impl` re-plans anyway. The plan and its stream sync still run only on a page-count change, so this costs nothing on steps where the page table is stable.

The recording rule moves into a small module-level helper (`_records_fa2_plan`) rather than staying an inline condition, so the part that regressed is directly unit-testable without a device — the surrounding `decode_plan()` closure needs FlashInfer and real KV buffers.

## Test Coverage

New CPU-only unit test: `tests/unittest/_torch/attention/test_flashinfer_plan_refresh.py`

- `test_every_non_trtllm_gen_decode_plan_is_recorded_under_graphs` / `test_trtllm_gen_is_not_recorded` / `test_eager_plans_are_not_recorded` cover the recording rule, including the wrapper class that regressed (graph-captured, backend resolved to `'auto'`, no tensor-core wrapper) which the old condition dropped.
- `test_page_table_change_replans_every_recorded_wrapper` covers the multi-wrapper case: two plan-params entries both carrying a page tuple must both be re-planned when the generation page counts change.
- `test_an_unchanged_page_table_does_not_replan` and `test_no_generations_clears_the_recorded_page_tuple` pin the two no-op paths.

End-to-end, the change was validated on a mixed sliding-window/full-attention MoE checkpoint on 4 GPUs (TP4/EP4) with CUDA graphs and the overlap scheduler: a 16-example GSM8K canary goes from 18.8 to 87.5 and a 100-example set from 7.0 to 95.0, matching the reference implementation. A fully cached-prefix decode reproducer that produced non-finite logits at the first decode step on 10/10 runs now produces none. Eager (no-graph) results are unchanged.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## Dev Engineer Review

- The stated functional change is to record page-count tuples for non-`trtllm-gen` decode wrappers during CUDA-graph capture.
- The supplied diff shows only import-formatting changes. It does not show the stated scheduling logic change.
- Confirm that the intended source changes are included in the PR before merge.
- No configuration or test-list files changed.

## QA Engineer Review

- `tests/unittest/_torch/attention/test_flashinfer_attention.py` is modified.
- The supplied diff shows only import-formatting changes. It does not show the stated new CUDA-graph test coverage.
- Confirm that the page-count recording, wrapper re-planning, unchanged-page, and no-generation tests are present and listed in the applicable CI or QA test lists.
- Verdict: needs follow-up.

### CODE REVIEW

File: tensorrt_llm/_torch/attention/backends/flashinfer.py [hunk 1]
```diff
@@ -2109,7 +2125,12 @@ def decode_plan():
                 q_len_per_req=plan_params.q_len_per_req,
                 disable_split_kv=False,
             )
-            if use_graph_tensor_cores and decode_wrapper._backend == 'fa2':
+            # Without this, multi-wrapper (variable sliding window) metadata
+            # deferred the re-plan to forward_impl and replayed the
+            # capture-time plan (one max_seq_len dummy in generation slot 0,
+            # one-token dummies elsewhere), so slots >= 1 attended only their
+            # first kv_chunk_size tokens.
+            if _records_fa2_plan(self.is_cuda_graph, decode_wrapper._backend):
```
Comment (lfr-0531) [inline tensorrt_llm/_torch/attention/backends/flashinfer.py]: Nit: it seems that we don't need to add a new `_records_fa2_plan` here, and we can just change the `use_graph_tensor_cores` to `self.is_cuda_graph` and `decode_wrapper._backend == 'fa2'` to `decode_backend != 'trtllm-gen'`. And please also simplify these comments.

```
if self.is_cuda_graph and decode_wrapper._backend != 'trtllm-gen':
```

File: tests/unittest/_torch/attention/test_flashinfer_plan_refresh.py
Comment (lfr-0531) [inline tests/unittest/_torch/attention/test_flashinfer_plan_refresh.py:1]: Please add these tests to `test_flashinfer_attention.py`.

**Review by lfr-0531 — APPROVED** [review]

File: tensorrt_llm/_torch/attention/backends/flashinfer.py (hunk 1 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py]: Done — inlined the condition as suggested (dropped `_records_fa2_plan`, now `if self.is_cuda_graph and decode_wrapper._backend != 'trtllm-gen':`) and trimmed the comment. One note on equivalence: `!= 'trtllm-gen'` is not identical to `== 'fa2'`. On non-SM90 (e.g. Blackwell) a 'fa2'-requested decode wrapper is built with `backend='auto'`, which flashinfer resolves during `plan()` to a concrete backend that can be 'fa3', so `== 'fa2'` would skip recording it and the multi-wrapper CUDA-graph re-plan would not fire; `!= 'trtllm-gen'` is the right dichotomy (only trtllm-gen self-refreshes each step), matching the existing `!= "trtllm-gen"` checks at `flashinfer.py:352` and `:2094`.

File: tests/unittest/_torch/attention/test_flashinfer_plan_refresh.py
Comment (brnguyen2) [reply tests/unittest/_torch/attention/test_flashinfer_plan_refresh.py:1]: Moved the plan-refresh tests into `test_flashinfer_attention.py` and deleted this standalone file. The three `_refresh_fa2_cuda_graph_plans` behavior tests moved over unchanged (validated: 3 passed). The three unit tests that called `_records_fa2_plan` directly were dropped since that helper is now inlined and gone — after inlining they would just re-assert the boolean, not cover the call site. Happy to add a small parametrized test of the inline gate (`self.is_cuda_graph and _backend != 'trtllm-gen'`) if you would prefer to keep that truth table explicit.

**Review by lfr-0531 — APPROVED** [review]

**Review by yuxianq — APPROVED** [review]
